pkgs/by-name/gh/ghostty/package.nix

```diff
@@ -0,0 +1,149 @@
+{
+  lib,
+  stdenv,
+  bzip2,
+  callPackage,
+  expat,
+  fetchFromGitHub,
+  fontconfig,
+  freetype,
+  git,
+  glib,
+  gsettings-desktop-schemas,
+  gtk4,
+  harfbuzz,
+  libadwaita,
+  libGL,
+  libpng,
+  libX11,
+  libXcursor,
+  libXi,
+  libXrandr,
+  ncurses,
+  nix-update-script,
+  nixosTests,
+  oniguruma,
+  pandoc,
+  pkg-config,
+  removeReferencesTo,
+  versionCheckHook,
+  wrapGAppsHook4,
+  zig_0_13,
+  zlib,
+}:
+let
+  # Ghostty needs to be built with --release=fast, --release=debug and
+  # --release=safe enable too many runtime safety checks.
+  zig_hook = zig_0_13.hook.overrideAttrs {
+    zig_default_flags = "-Dcpu=baseline -Doptimize=ReleaseFast --color off";
+  };
```

Comment:
_This is a question for the community, not a request:_ Might it be worth having separate fast, debug, and safe packages, possibly with a `release ? "fast" # or "debug", or "safe"` parameter? I guess the debug build might be useful for Ghostty users who want to file detailed bug reports, and from the [docs](https://github.com/Passw/ghostty-org-ghostty/blob/a8e5eef11cc67f87f445626f9ca2993373774bf8/PACKAGING.md#build-options) (my emphasis), it looks like the safe build would appeal to some users:

> `-Doptimize=ReleaseFast`: Build with optimizations enabled and safety checks disabled. This is the recommended build mode for distribution. **I'd prefer a safe build** but terminal emulators are performance-sensitive and the safe build is currently too slow. I plan to improve this in the future. Other build modes are available: `Debug`, `ReleaseSafe`, and `ReleaseSmall`.

Reply:
I think this would be something best left to a follow up PR that exposes more build options in general (like the rendering backend, font backend, etc.). I like it though ;)

Reply:
The debug builds are intended only for developers and are *insanely* slow (like, slower by *a hundredfold*, and there's a banner warning you that it's in debug mode and performance is going to be absolutely terrible). 

The merit of offering a ReleaseSafe build is debatable but currently there's a lot of sanity checks in ReleaseSafe that also grind performance to a halt. Maybe we should disable some of them for ReleaseSafe, but that's a decision for later.

TL;DR ReleaseFast is the only option as of now that offers any reasonable performance.

Reply:
https://github.com/ghostty-org/ghostty/blob/main/nix/package.nix#L28-L39 is upstream. (possible follow up pr...)

Reply:
_Both_ ReleaseSafe and Debug builds are very very slow due to additional runtime checks. The only people that should be using it are people developing Ghostty and those people should have a local development environment set up in a git checkout (Ghostty upstream includes a flake.nix with a devshell). In addition, if you're experimenting with non-default backends and the like you should also be building from a git checkout. This package is intended for general users and giving them footguns is just going to cause unnecessary pain.

Reply:
Speaking for the Zig project, the intention is for the upstream application developer to get first choice of the optimization mode. The Ghostty project does not choose one, however, since it uses this line:

https://github.com/ghostty-org/ghostty/blob/v1.0.0/build.zig#L49

This means that it falls to the distribution to choose the preferred optimization mode, and NixOS should probably make a default Zig policy that covers applications which do not populate [the preferred_optimize_mode field](https://ziglang.org/documentation/0.13.0/std/#std.Build.StandardOptimizeOptionOptions).

If Ghostty contributors want to prefer a specific optimization mode, then I suggest to edit that line in build.zig, communicating which optimization mode you want.

Reply:
> I strongly suggest for either the upstream application developer

[Upstream said ReleaseSafe ~2 months ago](https://lobste.rs/s/ineb98/ghostty_1_0_is_coming#c_ro3yws).

> Both ReleaseSafe and Debug builds are very very slow due to additional runtime checks.

That's true for Debug, but please, whoever is reading this, don't take this as a general advise for `ReleaseSafe`. Perf impact highly depends on the application; it may as well be negligible or not matter at all (not the case here though, according to Mitchell).

E.g. an old anecdote of myself looking into the Zig distribution: https://github.com/ziglang/zig/issues/15194#issuecomment-1511228015

Reply:
> [Upstream said ReleaseSafe ~2 months ago](https://lobste.rs/s/ineb98/ghostty_1_0_is_coming#c_ro3yws).

The linked comment says:

> For memory safety, we currently recommend **ReleaseFast**